core/src/test/scala/integration/kafka/api/MultiClusterIntegrationTestHarness.scala

```diff
@@ -46,27 +46,25 @@ abstract class MultiClusterIntegrationTestHarness extends MultiClusterKafkaServe
   val producerConfigs: Buffer[Properties] = new ArrayBuffer(numClusters) // referenced in MultiClusterAbstractConsumerTest
   val consumerConfigs: Buffer[Properties] = new ArrayBuffer(numClusters) // ditto
   val adminClientConfigs: Buffer[Properties] = new ArrayBuffer(numClusters)
-  val serverConfig = new Properties // TODO: should this be per-cluster, too? (seems likely)
+  val serverConfig = new Properties // unconditional extra configs for all brokers
 
   private val consumers = mutable.Buffer[KafkaConsumer[_, _]]()
   private val producers = mutable.Buffer[KafkaProducer[_, _]]()
   private val adminClients = mutable.Buffer[Admin]()
 
-  protected def interBrokerListenerName: ListenerName = listenerName
-
-  protected def modifyConfigs(props: Seq[Properties]): Unit = {
-    configureListeners(props)
-    props.foreach(_ ++= serverConfig)
-  }
-
-  override def generateConfigsByCluster(clusterId: Int): Seq[KafkaConfig] = {
-    val cfgs = TestUtils.createBrokerConfigs(brokerCountPerCluster, zkConnect(clusterId),
+  override def generateConfigsByCluster(clusterIndex: Int): Seq[KafkaConfig] = {
+    val cfgs = TestUtils.createBrokerConfigs(brokerCountPerCluster, zkConnect(clusterIndex),
       interBrokerSecurityProtocol = Some(securityProtocol), trustStoreFile = trustStoreFile,
       saslProperties = serverSaslProperties, logDirCount = logDirCount)
-    modifyConfigs(cfgs)
+    modifyConfigs(cfgs, clusterIndex)
     cfgs.map(KafkaConfig.fromProps)
   }
 
+  protected def modifyConfigs(props: Seq[Properties], clusterIndex: Int): Unit = {
```

Comment:
The clusterIndex is not used in this method. Can we remove it for now?

Reply:
This is a test harness, and the whole point of protected (or public) methods is that subclasses can override them--in particular, test-case subclasses.  The default implementation has no need for cluster-specific config modifications, but upcoming tests absolutely will.  Here's my WIP from the prototype:

```
+  // Set up brokers and controllers as follows:
+  //    +---------------------------------------+    +---------------------------------------+
+  //    | PHYSICAL CLUSTER(-INDEX) 0            |    | PHYSICAL CLUSTER(-INDEX) 1            |
+  //    |  - broker 100 = preferred controller  |    |  - broker 200 = preferred controller  |
+  //    |  - broker 101 = preferred controller  |    |  - broker 201 = preferred controller  |
+  //    |  - broker 102 = data broker           |    |  - broker 202 = data broker           |
+  //    |  - broker 103 = data broker           |    |  - broker 203 = data broker           |
+  //    |  - broker 104 = data broker           |    |  - broker 204 = data broker           |
+  //    +---------------------------------------+    +---------------------------------------+
+  // The bootstrap-servers list for each cluster will contain all five brokers.
+  override def modifyConfigs(props: Seq[Properties], clusterIndex: Int): Unit = {
+    debug(s"GRR DEBUG:  beginning ProxyBasedFederationTest modifyConfigs() override for clusterIndex=${clusterIndex}")
+    super.modifyConfigs(props, clusterIndex)
+    (0 until brokerCountPerCluster).map { brokerIndex =>
+      // 100-104, 200-204
+      val brokerId = 100*(clusterIndex + 1) + brokerIndex
+      debug(s"GRR DEBUG:  clusterIndex=${clusterIndex}, brokerIndex=${brokerIndex}: setting broker.id=${brokerId}")
+      props(brokerIndex).setProperty(KafkaConfig.BrokerIdProp, brokerId.toString)
+      if (brokerIndex < 2) {
+        // true for brokerIds 100, 101, 200, 201 only
+        debug(s"GRR DEBUG:  clusterIndex=${clusterIndex}, brokerIndex=${brokerIndex}, broker.id=${brokerId}: setting preferred.controller=true")
+        props(brokerIndex).setProperty(KafkaConfig.PreferredControllerProp, "true")
+      } else {
+        debug(s"GRR DEBUG:  clusterIndex=${clusterIndex}, brokerIndex=${brokerIndex}, broker.id=${brokerId}: leaving preferred.controller=false")
+      }
+    }
+    debug(s"GRR DEBUG:  done with ProxyBasedFederationTest modifyConfigs() override for clusterIndex=${clusterIndex}\n\n\n\n\n")
+  }
```

Reply:
Thanks for the explanation.